### DISCUSSION TITLE
What happens everynight at midnight to postgresql?

### DISCUSSION THREAD

**Discussion by dharapvj** [immich-app/immich#7108] [Q&A] [answered]
Every night at precise 12am my immich postgres instance "wakes up" and then works rigorously for about 1.5 hours!

I wonder why and what goes on at that time and why that process takes that a long time. There are no log entries in the container log. Can someone care to elaborate?

![image](https://github.com/immich-app/immich/assets/1920926/ec0d9f0c-702a-4ea9-b94c-9495c385b173)

**Comment by bo0tzz** [comment]
Immich runs a few housekeeping tasks every midnight, but as far as I'm aware none of those should be loading postgres for 1.5 hours straight. Can you try setting Immich's log level to `verbose` to see if that gives any clues, and can you give a bit more detail about your instance like how many assets you have and such? Also, is the load only on postgres or are any of the Immich containers also showing a higher load?

**Reply by dharapvj** [reply]
So.. after verbose logging.. I see that there is face recognition related batch that is going on... for 1.5 hours.. But.. ideally there is no need of any face recognition as already all faces have been recognized!

How do I stop this unnecessary, heat-generating, noise emitting, nightly exercise?

![image](https://github.com/immich-app/immich/assets/1920926/f157860f-bec9-4b16-b942-a85d032786b6)

**Reply by dharapvj** [reply]
CPU Load is shown only on PG DB. Immich micro-service I think "waits" while PG container toils.

Stats as shown in screenshot:

![image](https://github.com/immich-app/immich/assets/1920926/3e2275bd-c398-4108-8fb2-6ca955d84102)

**Reply by bo0tzz** [reply]
@mertalev any ideas? Is this because of the recognition distance causing some faces to not get matched and thus reprocessed?

**Reply by mertalev** [reply]
This is because of the change in #6634. The clustering we do at upload is much more limited now, so the nightly job is needed to make it work. But it isn't necessary if there were no faces detected since the last job.

**Comment by 4f1sh3r** [comment]
I am seeing a similar behaviour, but I am not 100% sure if its the same root cause. 

On my installation, facial recognition seems to re-run every night, but it seems not to (re)process all pictures. Only about 10.000 pictures seems to be reprocessed, of 100.000 in total. 

I only noticed it because of the high database load every morning, which lasts until every afternoon. Facial Recognition seems to run exclusively in the database. In my case the other containers have no load at all, it is completely limited to the database. 

By chance, I then saw in the jobs in the morning that there were several thousand facial recognition images in the queue. Actually, facial recognition has long finished for all my images and all relevant images have been recognized, so I don't understand why this is happening.

Can you check whether the same thing is happening to you?
I have not found any relevant log entries either.

**Comment by dharapvj** [comment]
So how do we get this fixed? Is there a code fix required?

How do I pause this nightly job for now?

**Reply by mertalev** [reply]
You can set "min recognized faces" to 1.

**Reply by mertalev** [reply]
To clarify a bit more, setting this will make the next facial recognition job assign or create a person for every face. Since the nightly job queues faces with no associated person, the job will no longer do anything.

**Reply by dharapvj** [reply]
Hi... can you please help me understand where shall I set this? Also - I could not understand impact of making this change on overall face recognition behavior.

Lastly, how can we debug more on what is happening and fix the real issue? I really am a happy user of Immich but due to this 2 hour nightly hissing.. I have shutdown my immich instance completely. Please help resolve this postgresql behavior. Can you move this discusison as issue and fix it?

**Reply by mertalev** [reply, marked as answer]
What version of Immich are you using? Facial recognition is much faster on the current release. Also, the advice to set it to 1 is outdated: setting it to 2 or 3 is optimal for speed (the default is 3).

**Reply by dharapvj** [reply]
Ok I have upgraded to latest version as of now. I will see how the jobs run at night.

**Reply by theone11** [reply]
> What version of Immich are you using? Facial recognition is much faster on the current release. Also, the advice to set it to 1 is outdated: setting it to 2 or 3 is optimal for speed (the default is 3).

I have the latest version 1.105.1 and it does not allow me change the concurrency of the Facial Recognition.
It is set at "1" and when I hover above it th emouse changes to a red circle with red line through it
![image](https://github.com/immich-app/immich/assets/3348953/de7cbda2-e5db-47d9-9d28-ea15c8a48e2a)

I would also like to understand why every night (midnight) I get >50000 waiting to be processed for a database that hasen't changed much for the last 2 months (only daily 10-20 pictures added)

**Comment by dharapvj** [comment]
With latest version of immich the postgres is behaving fine. the job ends in 4-5 minutes.
